shared/js/dataDriver.js

```diff
@@ -0,0 +1,283 @@
+        }
+      }
+      for (let i = 0; i < 12; i++) {
+        id += CHARS.charAt(lastRandChars[i]);
+      }
+      if (id.length !== 20) {
+        throw new Error("Length should be 20.");
+      }
+
+      return id;
+    };
+  }();
+
+  return class DataDriver {
+    /** ************ **
+     *** DataDriver ***
+     ** ************ **/
+
+    /**
+     * Create a DataDriver and initialize server connections.
+     */
+    constructor(options) {
+      if (!options.dispatcher) {
+        throw new Error("Missing option dispatcher");
+      }
+
+      this._dispatcher = options.dispatcher;
+      this._dispatcher.register(this, [
+        "fetchServerData",
+        "setupWindowData"
+      ]);
+    }
+
+    /**
+     * Send a text chat message by storing in the database.
+     */
+    sendTextChatMessage(message) {
+      this.update("chat", this.makeId(), message);
+    }
+
+    /** **************** **
+     *** Action Handler ***
+     ** **************** **/
+
+    /**
+     * Handle fetchServerData action by connecting to the room.
+     */
+    fetchServerData({ token }) {
+      this._connectToRoom(token);
+    }
+
+    /**
+     * Handle setupWindowData action by connecting to the room.
+     */
+    setupWindowData({ roomToken }) {
+      this._connectToRoom(roomToken);
+    }
+
+    /** *************** **
+     *** EventListener ***
+     ** *************** **/
+
+    /**
+     * Handle various events from addEventListener.
+     */
+    handleEvent(event) {
+      switch (event.type) {
+        case "open":
+          this._dispatcher.dispatch(new actions.DataChannelsAvailable({
+            available: true
+          }));
+          break;
+
+        case "put": {
+          // Ignore initial put from connecting.
+          let { data, path } = JSON.parse(event.data);
+          if (path === "/") {
+            return;
+          }
+
+          // Remove the leading "/" from the path.
+          this._processRecord(path.slice(1), data);
+          break;
+        }
+      }
+    }
+
+    /** ************* **
+     *** Data Source ***
+     ** ************* **/
+
+    /**
+     * Get the base url for the Firebase connection.
+     */
+    get BASE_FIREBASE() {
+      return "https://blinding-fire-8842.firebaseio.com/loop-test";
+    }
+
+    /**
+     * Get the maximum query limit value: 2^31 - 1.
+     */
+    get MAX_LIMIT() {
+      return 2147483647;
+    }
+
+    /**
+     * Get the server time for a local time or now.
+     */
+    getServerTime(time = Date.now()) {
+      return (this._clockSkew || 0) + time;
+    }
+
+    /**
+     * Generate an id based on the time.
+     */
+    makeId(time = this.getServerTime()) {
+      // Call the function that has locally scoped state.
+      return makeId(time);
+    }
+
+    /**
+     * Request the most recent chat messages between two times.
+     */
+    requestChat(startTime = 0, endTime = this.getServerTime(), limit = 0) {
```

Comment:
general nit: we also need to add return type documentation. It wasn't clear that this was returning a promise.